homeassistant/components/connectedcars/config_flow.py

```diff
@@ -0,0 +1,143 @@
+"""Config flow for ConnectedCars.io integration."""
+import logging
+
+from connectedcars.client import ConnectedCarsClient
+from connectedcars.constants import QUERY_USER, QUERY_VEHICLE_VIN
+from connectedcars.exceptions import ConnectedCarsException
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+
+from .const import DEFAULT_NAMESPACE, DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required("namespace", default=DEFAULT_NAMESPACE): str,
+        vol.Required("username"): str,
+        vol.Required("password"): str,
+        vol.Optional("vin"): str,
+    }
+)
+
+
+class ConnectedcarsApiHandler:
+    """Connected Cars.io API connector class."""
+
+    def __init__(self, namespace):
+        """Initialize."""
+        self.namespace = namespace
+        self.vin = None
+        self.client = None
+        self.user_data = None
+
+    async def authenticate(self, username, password) -> bool:
+        """Test if we can authenticate with the namespace."""
+        client = ConnectedCarsClient(username, password, self.namespace)
+
+        self.client = client
+
+        try:
+            self.user_data = await client.async_query(QUERY_USER)
+        except ConnectedCarsException:
+            return False
+
+        return True
+
+    async def get_email(self) -> str:
+        """Will get the email address of the account holder."""
+        if not self.user_data:
+            self.user_data = await self.client.async_query(QUERY_USER)
+
+        email = self.user_data["data"]["viewer"]["email"]
+
+        return email
+
+    async def get_vin(self, userinput_vin) -> str:
+        """Will get the vin identifier of the car."""
+        if not self.vin:
+            response = await self.client.async_query(QUERY_VEHICLE_VIN)
+            # This needs to lookup the "userinput_vin" to see if it can be found in the result, not just take the first car
+            vin = response["data"]["viewer"]["vehicles"][0]["vehicle"]["vin"]
+            self.vin = vin
+
+        return self.vin
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    ccah = ConnectedcarsApiHandler(data["namespace"])
+
+    if not await ccah.authenticate(data["username"], data["password"]):
+        raise InvalidAuth
+
+    try:
+        email = await ccah.get_email()
+    except ConnectedCarsException as exception:
```

Comment:
It would be nice if the upstream package already could throw more specific errors, remove the need to handle this all individually here.

Reply:
I did not create the upstream package, and I therefore do not have control over it, but I do agree.